pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+
+# Caching
+> There are only two hard problems in computer science: Naming things, cache invalidation and off by one errors.
+
+Indexing the entire project, its dependencies, elixir and erlang when
+the language server starts is likely going to be too performance
+intensive to do (to wit, rewriting atoms during namespacing takes 15
+seconds on our moderately sized project), so we will need some sort of
+on-disk representation of the index.
+
+Such a representation will need to take the following into account:
+
+  * Branches can while the language server is running, invalidating large portions of the cache
+  * Project level search and replace can affect large numbers of files at once
+  * The versions of elixir and erlang can change via `asdf` whenever the user wants
+
+Basicallly, the project's assets can change at any time, and it's not
+guaranteed that the language server will be able to detect all of the
+changes, so the language server needs to treat this cache with some
+skepticism and constantly "check its work".
+
+## Cache layout
+A cache entry will have the following format:
+
+```elixir
+
+@type source :: Lexical.uri()
+@entry_type :: :otp | :elixir | :dependency | :project
+
+@type entry %{
+  type: entry_type(),
+  source: source(),
+  entry_version: String.t(),
+  added_at: DateTime.t(),
+  value: term()
+}
+```
+
+### Invalidation
+When lexical starts the first time, it will need to index the
+following entities in the following order:
+
+  1. The elixir release
+  2. All project dependencies
+  3. The project's source code and built artifacts
+
+The index needs to take into account which stage added it, and include an appropriate
+version when adding an entry. The versions are as follows:
+
+| Entry Type  | Entry Version |
+| ----------- | --------------|
+| :otp        | [This insanity](https://stackoverflow.com/questions/9560815/how-to-get-erlangs-release-version-number-from-a-shell/34326368#34326368)|
+| :elixir     | `System.version()`|
+| :dependency | `Application.spec(dep_atom)[:vsn]`|
+| :project    | `md5(file contents)` |
+
+Entries in the cache should be replaced and not updated, and we should
+have a mechanism to sweep old entries from the cache (for example, if
+a dependecy version is old and hasn't been updated in two weeks, we
+should remove it). This invalidation is going to be one of the tricky
+parts of the cache, and we should be aggressive with it, as
+re-indexing things is likely to be less bad than having an
+ever-growing cache on disk.  We need to think more about this and come
+up with something that strikes a balance between efficiency and
+on-disk storage.
+
+# Indexing
+Indexing and searching go hand-in-hand, you need to make an index in a
+format that's useful, and as such, we'll approach this at a later
+date. In the interim, we can probably get away with a linear search;
+there aren't _that many_ things that we're indexing and elixir is kind
+of fast for this kind of thing.
```

Comment:
the index is going to go on disk, right?

https://www.erlang.org/doc/man/erlang.html#term_to_binary-1 for now?

Reply:
Yes, I was considering using something like `disk_log` or `dets` for this. Both use `term_to_binary` under the hood